pkg/httpserver/server_test.go

```diff
@@ -0,0 +1,185 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package httpserver
+
+import (
+	"context"
+	"crypto"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"net"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap/zaptest"
+
+	"storj.io/storj/pkg/pkcrypto"
+)
+
+var (
+	testKey = mustSignerFromPEM(`-----BEGIN PRIVATE KEY-----
+MIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQgT8yIof+3qG3wQzXf
+eAOcuTgWmgqXRnHVwKJl2g1pCb2hRANCAARWxVAPyT1BRs2hqiDuHlPXr1kVDXuw
+7/a1USmgsVWiZ0W3JopcTbTMhvMZk+2MKqtWcc3gHF4vRDnHTeQl4lsx
+-----END PRIVATE KEY-----
+`)
+	testCert = mustCreateLocalhostCert()
+)
+
+func TestServer(t *testing.T) {
+	address := "localhost:0"
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprint(w, "OK")
+	})
+	tlsConfig := &tls.Config{
+		Certificates: []tls.Certificate{
+			{
+				Certificate: [][]byte{testCert.Raw},
+				PrivateKey:  testKey,
+			},
+		},
+	}
+
+	testCases := []serverTestCase{
+		{
+			Name:    "missing address",
+			Handler: handler,
+			NewErr:  "server address is required",
+		},
+		{
+			Name:    "bad address",
+			Address: "this is no good",
+			Handler: handler,
+			NewErr:  "unable to listen on this is no good: listen tcp: address this is no good: missing port in address",
+		},
+		{
+			Name:    "missing handler",
+			Address: address,
+			NewErr:  "server handler is required",
+		},
+		{
+			Name:    "success via HTTP",
+			Address: address,
+			Handler: handler,
+		},
+		{
+			Name:      "success via HTTPS",
+			Address:   address,
+			Handler:   handler,
+			TLSConfig: tlsConfig,
+		},
+	}
+
+	for _, testCase := range testCases {
+		testCase := testCase
+		t.Run(testCase.Name, func(t *testing.T) {
+			s, ok := testCase.NewServer(t)
+			if !ok {
+				return
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			errCh := make(chan error, 1)
+			go func() {
+				errCh <- s.Run(ctx)
+			}()
+
+			testCase.DoGet(t, s)
+
+			cancel()
+			select {
+			case err := <-errCh:
+				assert.NoError(t, err, "Run() should not have failed")
+			case <-time.After(time.Minute):
+				assert.Fail(t, "timed out waiting for Run() to return")
+			}
+		})
+	}
+}
+
+type serverTestCase struct {
+	Name      string
+	Address   string
+	Handler   http.Handler
+	TLSConfig *tls.Config
+	NewErr    string
+}
+
+func (c *serverTestCase) NewServer(tb testing.TB) (*Server, bool) {
```

Comment:
As before, the name of the receiver should be larger.

Reply:
:+1: